### PR TITLE
fix OpenGL version detection

### PR DESCRIPTION
**Pull request by Be-ing**
This allows OpenGL waveforms to be used with the Wayland Qt Platform Abstraction on KDE and Weston. Though GL_MAJOR_VERSION and GL_MINOR_VERSION reported much higher versions (4.6 when tested on Intel UHD Graphics 770 (ADL-S GT1)), QSurfaceFormat reports 2.0 on KDE Wayland and Weston (somehow it reports 2.1 on GNOME Wayland though) and 2.1 with the Xcb QPA. QSurfaceFormat does not report higher versions if they are requested in the OpenGLWindow constructor.

This has only been tested on Linux. @m0dB can you test on macOS? @JoergAtGithub could you test on Windows?

### CODE REVIEW

**Comment by Be-ing** [comment]
Finally with Wayland high DPI scaling Just Works automatically without needing to set the environment variable `QT_AUTO_SCREEN_SCALE_FACTOR=1`.

File: src/waveform/waveformwidgetfactory.cpp [hunk 1]
```diff
@@ -142,13 +142,29 @@ WaveformWidgetFactory::WaveformWidgetFactory()
...
+            qDebug().noquote() << QLatin1String(
+                    "OpenGL driver version string \"%1\", vendor \"%2\", "
+                    "renderer \"%3\"")
+                                          .arg(versionString, vendorString, rendererString);
+
+            GLint majorVersion = -1;
+            glFunctions->glGetIntegerv(GL_MAJOR_VERSION, &majorVersion);
+            GLint minorVersion = -1;
+            glFunctions->glGetIntegerv(GL_MINOR_VERSION, &minorVersion);
+            if (majorVersion == -1 || minorVersion == -1) {
```
Comment (Be-ing) [inline src/waveform/waveformwidgetfactory.cpp:154]: If glGetIntegerv fails somehow I don't actually know what the value would be. I'm guessing it doesn't modify the passed in pointer, but I have no way to verify that.

**Comment by JoergAtGithub** [comment]
On my Windows11 laptop I get:
```
debug [Main] OpenGL driver version string "4.6.0 Compatibility Profile Context 22.20.28.15.230324", vendor "ATI Technologies Inc.", renderer "AMD Radeon(TM) Graphics"
debug [Main] Supported OpenGL version: 4.6
```

File: src/waveform/waveformwidgetfactory.cpp (hunk 1 above)
Comment (Swiftb0y) [reply src/waveform/waveformwidgetfactory.cpp:154]: I guess the proper thing to do would be to check for `GL_INVALID_ENUM` errors? https://registry.khronos.org/OpenGL-Refpages/gl4/html/glGet.xhtml#errors
Also apparently you can rely on the `GL_VERSION` string to extract the version `GL_MAJOR/MINOR_VERSION` is not available...
https://www.khronos.org/opengl/wiki/OpenGL_Context#Context_information_queries

File: src/waveform/waveformwidgetfactory.cpp [hunk 2]
```diff
@@ -142,13 +142,29 @@ WaveformWidgetFactory::WaveformWidgetFactory()
...
                     reinterpret_cast<const char*>(glFunctions->glGetString(GL_RENDERER))));
-            qDebug() << versionString << vendorString << rendererString;
-
-            // note: the requested version has been set in WGLWidget's OpenGLWindow constructor
-            const int majorVersion = context->surface()->format().majorVersion();
-            const int minorVersion = context->surface()->format().minorVersion();
+            qDebug().noquote() << QStringLiteral(
+                    "OpenGL driver version string \"%1\", vendor \"%2\", "
+                    "renderer \"%3\"")
+                                          .arg(versionString, vendorString, rendererString);
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp:148]: That's what it was like before but the formatting was less than ideal:
`OpenGL driver version string "VERSION-STRING" vendor "VENDOR-STRING" renderer "RENDER-STRING"`

This way allows inserting the the commas between the fields:
`OpenGL driver version string "VERSION-STRING", vendor "VENDOR-STRING", renderer "RENDER-STRING"`

Using `<<` would insert awkward spaces before the commas:

``OpenGL driver version string "VERSION-STRING" , vendor "VENDOR-STRING" , renderer "RENDER-STRING"`

[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/waveform/waveformwidgetfactory.cpp:148]: Good point. don't really like that the resulting code is harder to read, but not worth arguing over.

File: src/waveform/waveformwidgetfactory.cpp (hunk 1 above)
Comment (Be-ing) [reply src/waveform/waveformwidgetfactory.cpp]: done

**Review by Swiftb0y — APPROVED** [review]
LGTM. can confirm this to also fix the issue on Fedora 38 (Gnome running XWayland). There are other issues on Wayland though I'll file in separate gh issues.

This probably doesn't make any difference on MacOS, but lets wait with merging until someone confirms it.

**Comment by m0dB** [comment]
I'll test on macOS this weekend.

**Comment by m0dB** [comment]
It seems to work fine on macOS. However, I wanted to check what the preferences dialogs shows... but the dialog doesn't open. Not caused by this PR, it also happens with 2.4 (after pulling and installing the new buildenv) . Is anybody else experiencing this? Any idea what may cause this?

**Comment by JoergAtGithub** [comment]
LGTM! Thank you!

**Pull request merged**